pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java

```diff
@@ -129,9 +135,14 @@ public static Optional<RawMessage> rebatchMessage(RawMessage msg,
                                                       msg.getMessageIdData().getPartition(),
                                                       i);
                 if (!singleMessageMetadata.hasPartitionKey()) {
-                    messagesRetained++;
-                    Commands.serializeSingleMessageInBatchWithPayload(singleMessageMetadata,
-                                                                      singleMessagePayload, batchBuffer);
+                    if (retainNullKey) {
```

Comment:
how about move this logic to filter?

Reply:
If `topicCompactionRemainNullKey=true` then we also need to retain `<null, "">` message but filter will can't retain that message with an empty payload.